File: LeetCode/Study_Plan/DataStructure/Python/53.py
#https://leetcode.com/problems/maximum-subarray/

#Kadane's Algorithm
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        curr_subarr  = max_subarr = nums[0]
        for n in nums[1:]:
            curr_subarr = max(n, curr_subarr + n)
            max_subarr = max(curr_subarr, max_subarr)
        return max_subarr

File: leetcode/study_plan/DataStructure/C_Plus_Plus/53.cpp
//https://leetcode.com/problems/maximum-subarray/

//Kadane's Algorithm

class Solution
{
    public:
        int maxSubArray(vector<int>& nums)
        {
            int curr_subarr =  nums[0];
            int max_subarr = nums[0];
            for ( int i = 1; i < nums.size(); i++ )
            {
                curr_subarr = max( nums[i], curr_subarr + nums[i] );
                max_subarr = max( curr_subarr, max_subarr );
            }
            return max_subarr;
        }
};